plugins/outputs/bigquery/bigquery.go

```diff
@@ -0,0 +1,229 @@
+package bigquery
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+const (
+	defaultOffSetKey   = "offset-key.json"
+	timeStampFieldName = "timestamp"
+)
+
+var defaultTimeout = 5 * time.Second
+
+const sampleConfig = `	
+  ## Credentials File
+  credentials_file = "/path/to/service/account/key.json"
+  ## Google Cloud Platform Project
+  project = "my-gcp-project"
+
+  ## The namespace for the metric descriptor
+  dataset = "telegraf"
+
+  ## Timeout for BigQuery operations.  Minimum timeout is "1s".
+  # timeout = "5s"
+`
+
+type BigQuery struct {
+	CredentialsFile string `toml:"credentials_file"`
+	Project         string `toml:"project"`
+	Dataset         string `toml:"dataset"`
+
+	Timeout internal.Duration `toml:"timeout"`
+
+	Log telegraf.Logger
+
+	client *bigquery.Client
+}
+
+// SampleConfig returns the formatted sample configuration for the plugin.
+func (s *BigQuery) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns the human-readable function definition of the plugin.
+func (s *BigQuery) Description() string {
+	return "Configuration for Google Cloud BigQuery to send entries"
+}
+
+func (b *BigQuery) Connect() error {
+	if b.Project == "" {
+		return fmt.Errorf("Project is a required field for BigQuery output")
+	}
+
+	if b.Dataset == "" {
+		return fmt.Errorf("Dataset is a required field for BigQuery output")
+	}
+
+	if b.Timeout.Duration < 1*time.Second {
+		b.Timeout.Duration = defaultTimeout
+	}
+
+	if b.client == nil {
+		return b.setUpDefaultClient()
+	}
+
+	return nil
+}
+
+func (b *BigQuery) setUpDefaultClient() error {
+	var credentialsOption option.ClientOption
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, b.Timeout.Duration)
+	defer cancel()
+
+	if b.CredentialsFile != "" {
+		credentialsOption = option.WithCredentialsFile(b.CredentialsFile)
+	} else {
+		creds, err := google.FindDefaultCredentials(ctx)
+		if err != nil {
+			return fmt.Errorf(
+				"unable to find Google Cloud Platform Application Default Credentials: %v."+
+					"Either set ADC or provide CredentialsFile config", err)
+		}
+		credentialsOption = option.WithCredentials(creds)
+	}
+
+	client, err := bigquery.NewClient(ctx, b.Project, credentialsOption)
+	b.client = client
+	return err
+}
+
+// Write the metrics to Google Cloud BigQuery.
+func (b *BigQuery) Write(metrics []telegraf.Metric) error {
+	groupedMetrics := b.groupByMetricName(metrics)
+
+	var wg sync.WaitGroup
+
+	for k, v := range groupedMetrics {
+		wg.Add(1)
+		go func(k string, v []bigquery.ValueSaver) {
+			defer wg.Done()
+			b.insertToTable(k, v)
+		}(k, v)
```

Comment:
This should also work
```suggestion
		go func() {
			defer wg.Done()
			b.insertToTable(k, v)
		}()
```